catroid/src/test/java/org/catrobat/catroid/test/visualplacement/VisualPlacementPositionCorrectionTest.java

```diff
@@ -0,0 +1,149 @@
+
+@RunWith(MockitoJUnitRunner.class)
+public class VisualPlacementPositionCorrectionTest {
+
+	@Mock
+	ImageView imageView;
+
+	@Mock
+	MotionEvent motionEvent;
+
+	@Mock
+	CoordinateInterface coordinateInterface;
+
+	@Mock
+	TimeProvider timeProvider;
+
+	@InjectMocks
+	VisualPlacementTouchListener listener;
+
+	private final Float initalPositionX = 1.f;
+	private final Float initalPositionY = 1.f;
+	private final Float firstMovementPositionX = 5.f;
+	private final Float firstMovementPositionY = 5.f;
+	private final Float secondMovementPositionX = 6.f;
+	private final Float secondMovementPositionY = 6.f;
+	private final long firstMovementTimestamp = 150;
+	private final long secondMovementTimestamp = 170;
+	private final long actionUpTimestamp = 175;
+
+	@Before
+	public void setUp() {
+		when(timeProvider.elapsedRealtime()).thenReturn((long) 0);
+	}
+
+	public void triggerTouchDownEvent(float startPositionX, float startPositionY) {
+		when(motionEvent.getAction()).thenReturn(MotionEvent.ACTION_DOWN);
+		when(imageView.getX()).thenReturn(startPositionX);
+		when(imageView.getY()).thenReturn(startPositionY);
+		listener.onTouch(imageView, motionEvent, coordinateInterface);
+	}
+
+	public void triggerMovementEvent(long timestamp, float positionX, float positionY) {
+		when(motionEvent.getAction()).thenReturn(MotionEvent.ACTION_MOVE);
+		when(motionEvent.getRawX()).thenReturn(positionX);
+		when(motionEvent.getRawY()).thenReturn(positionY);
+
+		when(timeProvider.elapsedRealtime()).thenReturn(timestamp);
+		when(imageView.getX()).thenReturn(positionX);
+		when(imageView.getY()).thenReturn(positionY);
+		listener.onTouch(imageView, motionEvent, coordinateInterface);
+	}
+
+	public void triggerTouchUpEvent(long timestamp) {
+		when(timeProvider.elapsedRealtime()).thenReturn(timestamp);
+		when(motionEvent.getAction()).thenReturn(MotionEvent.ACTION_UP);
+		listener.onTouch(imageView, motionEvent, coordinateInterface);
+	}
+
+	@Test
+	public void testTouchCorrection() {
+		triggerTouchDownEvent(initalPositionX, initalPositionY);
+		triggerMovementEvent(firstMovementTimestamp, firstMovementPositionX, firstMovementPositionY);
+		triggerMovementEvent(secondMovementTimestamp, secondMovementPositionX, secondMovementPositionY);
+		triggerTouchUpEvent(actionUpTimestamp);
+
+		verify(imageView).setX(firstMovementPositionX);
+		verify(imageView).setY(firstMovementPositionY);
+		verify(imageView, times(3)).setX(anyFloat());
+	}
+
+	@Test
+	public void testTouchCorrectionWithInvalidDelayBetweenMovements() {
+		long invalidActionUpTimestamp = secondMovementTimestamp + 15;
+
+		triggerTouchDownEvent(initalPositionX, initalPositionY);
+		triggerMovementEvent(firstMovementTimestamp, firstMovementPositionX, firstMovementPositionY);
+		triggerMovementEvent(secondMovementTimestamp, secondMovementPositionX, firstMovementPositionY);
+		triggerTouchUpEvent(invalidActionUpTimestamp);
+
+		verify(imageView, times(2)).setX(anyFloat());
```

Comment:
A better test would here check that the X and Y values at the end are the ones from the secondMovement. This makes it clearer for others who want to understand the test, and also is the functional end result. The number of invocations is OK too for this specific implementation and test case.

Reply:
the values get set onto the mock of CoordinateInterface using this code
`coordinateInterface.setXCoordinate(imageView.getX());` 
We already mock the getX() Method of the imageView mock before calling the method, inside it would be set to the right value with setX() but if I try to verify in the end it would still use the old value so I can't check it like that.

Reply:
Ah, OK, thanks!